core/storage/topic/gae_models.py

```diff
@@ -388,6 +383,15 @@ class TopicSummaryModel(base_models.BaseModel):
     # The thumbnail background color of the topic.
     thumbnail_bg_color = datastore_services.StringProperty(indexed=True)
     version = datastore_services.IntegerProperty(required=True)
+    # A mapping from each story id belonging to the topic to a list of
+    # exploration ids that are linked to the corresponding published story. This
+    # mapping represents the relationships between a topic, its published
+    # stories and the explorations that are linked to each story. This field
+    # must be kept in-sync with updates made to an owned story's explorations,
+    # to the topic's ownership of said stories, and to the publishing or
+    # unpublishing of said stories.
+    published_story_exploration_mapping = datastore_services.JsonProperty(
+        required=True)
```

Comment:
What's the consequences of putting `required=True` in an unsafe spot?

Reply:
That's what we'd need to test :) I'm concerned about GETs failing if a model has nothing populated in a required=True field.

Reply:
Did you test this?

Also, actually, you can probably do this test locally. Checkout the develop branch and generate some data. Then switch to your branch -- without killing the server -- and then try retrieving topics. See if you get any errors.

If you don't get any errors, that's a good sign and indicates that this is viable, but let's also do this test on a prod server to be safe (you can arrange this with @chris7716 who is one of the server admins).

Reply:
So I tested this locally by generating topics, stories, explorations, and translation suggestions on the `develop` branch. Then, I switched to the branch with all my current changes. Finally, I loaded reviewable translation opportunities as a translation reviewer on the CD dashboard. I don't think this is viable, as I got the following error:

```
ERROR:root:Exception raised at http://localhost:8181/getreviewableopportunitieshandler?topic_name=null: 'NoneType' object has no attribute 'values'
Traceback (most recent call last):
  File "/tmp/tmpn55mmd/lib/python3.8/site-packages/webapp2.py", line 604, in dispatch
    return method(*args, **kwargs)
  File "/open-source/oppia/core/controllers/acl_decorators.py", line 116, in test_can_access
    return handler(self, *args, **kwargs)
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 334, in get
    self._get_reviewable_exploration_opportunity_summaries(
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 378, in _get_reviewable_exploration_opportunity_summaries
    topic_services.get_all_published_story_exploration_ids())
  File "/open-source/oppia/core/domain/topic_services.py", line 1814, in get_all_published_story_exploration_ids
    return list(set(exp_ids))
  File "/open-source/oppia/core/domain/topic_services.py", line 1811, in <genexpr>
    summary.published_story_exploration_mapping.values())
AttributeError: 'NoneType' object has no attribute 'values'
ERROR:root:Exception raised: 'NoneType' object has no attribute 'values'
INFO     2024-02-05 22:50:11,946 module.py:883] default: "GET /getreviewablesuggestions/exploration/translate_content?offset=0&sort_key=Date&limit=20 HTTP/1.1" 200 79
Traceback (most recent call last):
  File "/tmp/tmpn55mmd/lib/python3.8/site-packages/webapp2.py", line 604, in dispatch
    return method(*args, **kwargs)
  File "/open-source/oppia/core/controllers/acl_decorators.py", line 116, in test_can_access
    return handler(self, *args, **kwargs)
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 334, in get
    self._get_reviewable_exploration_opportunity_summaries(
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 378, in _get_reviewable_exploration_opportunity_summaries
INFO     2024-02-05 22:50:11,948 module.py:883] default: "GET /getreviewableopportunitieshandler?topic_name=null HTTP/1.1" 500 81
    topic_services.get_all_published_story_exploration_ids())
  File "/open-source/oppia/core/domain/topic_services.py", line 1814, in get_all_published_story_exploration_ids
    return list(set(exp_ids))
  File "/open-source/oppia/core/domain/topic_services.py", line 1811, in <genexpr>
    summary.published_story_exploration_mapping.values())
AttributeError: 'NoneType' object has no attribute 'values'
ERROR:root:Exception raised at http://localhost:8181/getreviewableopportunitieshandler?topic_name=All: The supplied input topic: All is not valid
Traceback (most recent call last):
  File "/tmp/tmpn55mmd/lib/python3.8/site-packages/webapp2.py", line 604, in dispatch
    return method(*args, **kwargs)
  File "/open-source/oppia/core/controllers/acl_decorators.py", line 116, in test_can_access
    return handler(self, *args, **kwargs)
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 334, in get
    self._get_reviewable_exploration_opportunity_summaries(
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 382, in _get_reviewable_exploration_opportunity_summaries
    raise self.InvalidInputException(
core.controllers.base.UserFacingExceptions.InvalidInputException: The supplied input topic: All is not valid
ERROR:root:Exception raised: The supplied input topic: All is not valid
Traceback (most recent call last):
  File "/tmp/tmpn55mmd/lib/python3.8/site-packages/webapp2.py", line 604, in dispatch
    return method(*args, **kwargs)
  File "/open-source/oppia/core/controllers/acl_decorators.py", line 116, in test_can_access
    return handler(self, *args, **kwargs)
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 334, in get
    self._get_reviewable_exploration_opportunity_summaries(
  File "/open-source/oppia/core/controllers/contributor_dashboard.py", line 382, in _get_reviewable_exploration_opportunity_summaries
    raise self.InvalidInputException(
core.controllers.base.UserFacingExceptions.InvalidInputException: The supplied input topic: All is not valid

ERROR:root:Frontend error: 
Uncaught (in promise): Error: 'NoneType' object has no attribute 'values'
Error: 'NoneType' object has no attribute 'values'
    at eval (webpack:///./core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts?:146:23)
    at ZoneDelegate.invoke (webpack:///./node_modules/zone.js/dist/zone.js?:390:30)
    at Object.onInvoke (webpack:///./node_modules/@angular/core/fesm2015/core.js?:29022:33)
    at ZoneDelegate.invoke (webpack:///./node_modules/zone.js/dist/zone.js?:389:36)
    at Zone.run (webpack:///./node_modules/zone.js/dist/zone.js?:147:47)
    at eval (webpack:///./node_modules/zone.js/dist/zone.js?:895:38)
    at ZoneDelegate.invokeTask (webpack:///./node_modules/zone.js/dist/zone.js?:425:35)
    at Object.onInvokeTask (webpack:///./node_modules/@angular/core/fesm2015/core.js?:29009:33)
    at ZoneDelegate.invokeTask (webpack:///./node_modules/zone.js/dist/zone.js?:424:40)
    at Zone.runTask (webpack:///./node_modules/zone.js/dist/zone.js?:192:51)

INFO     2024-02-05 22:50:12,014 module.py:883] default: "POST /frontend_errors HTTP/1.1" 200 85
    at URL: http://localhost:8181/contributor-dashboard
ERROR:root:Frontend error: 
Uncaught (in promise): Error: The supplied input topic: All is not valid
Error: The supplied input topic: All is not valid
    at eval (webpack:///./core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts?:146:23)
    at ZoneDelegate.invoke (webpack:///./node_modules/zone.js/dist/zone.js?:390:30)
    at Object.onInvoke (webpack:///./node_modules/@angular/core/fesm2015/core.js?:29022:33)
    at ZoneDelegate.invoke (webpack:///./node_modules/zone.js/dist/zone.js?:389:36)
    at Zone.run (webpack:///./node_modules/zone.js/dist/zone.js?:147:47)
    at eval (webpack:///./node_modules/zone.js/dist/zone.js?:895:38)
    at ZoneDelegate.invokeTask (webpack:///./node_modules/zone.js/dist/zone.js?:425:35)
    at Object.onInvokeTask (webpack:///./node_modules/@angular/core/fesm2015/core.js?:29009:33)
    at ZoneDelegate.invokeTask (webpack:///./node_modules/zone.js/dist/zone.js?:424:40)

```

Reply:
@seanlip Due to the failures that occurred locally, should I still try to get it tested on the prod server?

Reply:
@StephenYu2018 I was actually expecting a possible datastore-level error along the lines of "required field not set". It seems like that is not happening?

The error you have seems to be due to that field not being populated, which is something we should consider -- how do you plan to handle the `published_story_exploration_mapping = None` case gracefully? Because that's what is going to crop up if a topic is accessed before we do the regeneration.

Also because of this I think we should test this PR on a prod server before we merge it, to avoid a breakage in production. See https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production for details.

Reply:
Okay I added in defensive checks when getting published story exploration ids. When `published_story_exploration_mapping = None`, we use the topic summary id to fetch the corresponding topic, and we pass that topic into `compute_summary_of_topic()` to populate the `None` field, and replace the original topic summary with the resulting summary. I did the same procedure (setup data on develop branch, switch to issue branch, and go to contributions and review tab) and the reviewable opportunities are being fetched properly.

I still need to file a request with the server admins.

Reply:
Sounds good -- yes, please do file that request, we do need to ensure nothing breaks in prod. Please include the testing doc in your PR description.

I am a bit confused -- does compute_summary_of_topic() actually populate the field? If so then you might want to rename it compute_and_save_summary_of_topic() or similar (and verify that there isn't some other method already doing this). Otherwise it sounds like it only computes the summary and returns it, but doesn't persist it.

Also that means that required=True here isn't an issue (since it only applies on PUT), correct?